### DISCUSSION TITLE
Some of the single quotes in the synonyms column of the compound reference file are not escaped.

### DISCUSSION THREAD

**Discussion by jeevangelista** [nih-cfde/published-documentation#337] [Bug] [answered]
Example CID: 5337997

> ['Nifuroxazide', '965-52-6', 'Nifuroxazid', 'Dicoferin', 'Diarlidan', 'UNII-PM5LI0P38J', 'PM5LI0P38J', 'MLS000069620', 'Ercefuryl', 'BENZOIC ACID, p-HYDROXY-, (5-NITROFURFURYLIDENE)HYDRAZIDE', 'Nifuroxazida', 'Nifuroxazidum', '4-hydroxy-N-[(E)-(5-nitro-2-furyl)methyleneamino]benzamide', 'Ercefurol', 'RC-27109', 'SMR000058185', '4-hydroxy-N-[(E)-(5-nitrofuran-2-yl)methylideneamino]benzamide', **_'4-Hydroxy-N'-((5-nitrofuran-2-yl)methylene)benzohydrazide'_**, 'Benzoic acid, 4-hydroxy-, 2-[(5-nitro-2-furanyl)methylene]hydrazide', 'Nifuroxazidum [INN-Latin]', 'Nifuroxazida [INN-Spanish]', 'CCRIS 5490', 'RC 27109', 'p-Hydroxybenzoic acid (5-nitrofurfurylidene)hydrazide', 'R.C. 27-109', 'EINECS 213-522-0', 'BRN 1352180', 'Diaryl', 'Nifuroxazide [INN:BAN:DCF]', 'MFCD00079482', '4-Hydroxy-N'-(5-nitrofurfuryliden)benzhydrazid', 'Prestwick_874', '(Nitro-5' furfurylidene-2') hydroxy-4 benzhydrazide', 'Nifuroxazide (INN)', 'Opera_ID_40', '(Nitro-5' furfurylidene-2') hydroxy-4 benzhydrazide [French]', 'Prestwick2_000555', 'Prestwick3_000555', 'BSPBio_000629', 'SCHEMBL542542', 'SCHEMBL543480', 'p-Hydroxybenzoic acid 5-nitrofurfurylidene hydrazide', 'ARONIS016368', 'BPBio1_000693', 'CHEMBL244888', 'SCHEMBL14608895', 'CHEBI:94556', 'CHEBI:135136', 'HMS1569P11', 'HMS2096P11', 'HMS3713P11', 'Pharmakon1600-01505788', 'HY-B1436', 'ZINC7997719', 'BDBM50396177', 'NSC759261', 's4182', 'STK091243', 'STL454990', 'AKOS000483758', 'AC-8713', 'CCG-213989', 'CS-4918', 'DB13855', 'NSC 759261', 'NSC-759261', 'AS-13310', 'H868', 'BIM-0014420.P001', 'SBI-0014420.P002', 'AB00513861', 'SJ000287313', 'D07111', 'AB00513861_02', 'AB00513861_03', 'Nifuroxazide, VETRANAL(TM), analytical standard', '4-hydroxy-2'-(5-nitrofurfurylidene)benzohydrazide', '965N526', 'A845610', 'W-100139', 'BRD-K68188368-001-03-7', '4-Dydroxy-N-[(E)-(5-nitro-2-furyl)methyleneamino]benzamide', '4-hydroxy-N'-[(1E)-(5-nitro-2-furyl)methylene]benzohydrazide', 'Nifuroxazide, European Pharmacopoeia (EP) Reference Standard', '4-hydroxy-N'-[(E)-(5-nitrofuran-2-yl)methylidene]benzohydrazide', '4-hydroxy-N-[(E)-(5-nitrofuran-2-yl)methylidene]benzenecarbohydrazonic acid', 'Benzoic acid, 4-hydroxy-, [(1E)-(5-nitro-2-furanyl)methylene]hydrazide']

**Comment by abradyIGS** [comment]
Replicated; fixing and pushing an update, stand by

**Comment by abradyIGS** [comment]
This issue has now tested as "fixed" -- marking "answered," but please reopen and report if you find otherwise.

Ontology files and the submission prep script have all been updated and published [here](https://osf.io/bq6k9/): please overwrite local copies and try again.

Note: the master C2M2 JSON schema has also updated (to match previous development schema): it is [here](https://osf.io/c63aw/) and an up-to-date copy needs to be included with submissions during prep with the script.

**Reply by jeevangelista** [reply]
I'm still getting a CastError here:

[CastError('Field "synonyms" can\'t cast value "["26239-55-4", "N-(2-Acetamido)iminodiacetic acid", "2,2\'-((2-Amino-2-oxoethyl)azanediyl)diacetic acid", "ADA", "Glycine, N-(2-amino-2-oxoethyl)-N-(carboxymethyl)-", "Carbamoylmethylaminodiacetic acid", "2,2\'-[(2-amino-2-oxoethyl)imino]diacetic acid", "N-(Carbamoylmethyl)iminodiacetic acid", "N-Carbamoylmethyliminodi(acetic acid)", "MFCD00008031", "UNII-5C4R3O704E", "N-(2-Acetamido)-2-iminodiacetic acid", "(CARBAMOYLMETHYL-CARBOXYMETHYL-AMINO)-ACETIC ACID", "2-[(2-amino-2-oxoethyl)-(carboxymethyl)amino]acetic acid", "5C4R3O704E", "N-[2-Acetamido]iminodiacetic acid", "N-[2-ACETAMIDO]-2-IMINODIACETIC ACID", "ADA buffer", "EINECS 247-530-0", "a d a", "bmse000146", "N-(2-amino-2-oxoethyl)-N-(carboxymethyl)glycine", "SCHEMBL233033", "2-[(carbamoylmethyl)(carboxymethyl)amino]acetic acid", "ADA, >=98% (titration)", "DTXSID4067191", "CHEBI:43960", "n-(2-acetamide)iminodiacetic acid", "N-(2-Acetamido)imidodiacetic acid", "n-(-2-acetamido)iminodiacetic acid", "n-(2-acetamido)-iminodiacetic acid", "ZINC2545127", "BBL023210", "s6310", "STL355999", "ADA, BioXtra, >=98% (titration)", "AKOS005267103", "AKOS025395625", "AC-8316", "ADA, BioUltra, >=99.0% (T)", "DB02810", "MCULE-7776883622", "ADA, Vetec(TM) reagent grade, >=98%", "AS-12375", "SY048625", "A0699", "FT-0621892", "Iminodiacetic acid, N-(2-amino-2-oxoethyl)-", "D70536", "2,2\'-(2-amino-2-oxoethylazanediyl)diacetic acid", "2,2\\\'-[(2-amino-2-oxoethyl)imino]diacetic acid", "W-107186", "acetic acid, 2,2\'-[(2-amino-2-oxoethyl)imino]bis-", "Q27093765"]" for type "array" with format "default"')]

**Comment by abradyIGS** [comment]
Sorry -- we had other issues with single quotes being inappropriate delimiters for JSON strings, which was fixed, and I rolled my response above into the fix for that -- but what you quote above is a spec-compliant JSON string array, and none of our software is going to have any problem with it.

Single quotes are inconsistently escaped within the double-quoted names, but this is irrelevant: the only characters JSON requires escape sequences for inside double-quoted strings are the double-quote characters themselves. All other escape-like sequences are ignored by spec-compliant software and treated as string literals.

Note also that the single-quote-escape inconsistency is something we imported directly from the PubChem labels and is not an artifact of our processing. See [this page](https://pubchem.ncbi.nlm.nih.gov/compound/117765#section=Wikipedia), section 3.4.2, and search for a backslash: what you quote above, with one `\'` sequence in one of the listed synonyms, but with other (unescaped) single quotes present, matches the inconsistency in the synonyms as submitted by the depositor.

**Comment by abradyIGS** [comment]
[just so i can try to avoid falling on my face again this week -- i'm correct in thinking that CastError was not generated by CFDE software, right? if i'm wrong and it was, please let me know which bit so i can go yell at it]

**Comment by jeremyjyang** [comment]
Today I get this cfde-submit error which seems relevant to this discussion:

> [DEBUG] cfde_submit.validation::validate_user_submission() Validating TableSchema in BDBag '/home/data/CFDE/data/drugpages2021/submission_output.zip'
> Validation error in compound.tsv
> Type error in the cell "["Chloropyramine", "Halopyramine", "Chlorpyramine", "Anaphylline", "59-32-5", "Suprastin", "Avapena", "Synopen", "Synpen", "Chlorneoantergan", "Chloropyribenzamine", "Chlorpyraminum", "Allergan S", "Synopen R", "Chloroneoantergan", "Chloropyramine [INN:BAN:DCF]", "UNII-2K3L8O9SOV", "NSC 241038", "N-(p-Chlorobenzyl)-N',N'-dimethyl-N-(2-pyridyl)ethylenediamine", "2K3L8O9SOV", "2-((p-Chlorobenzyl)(2-(dimethylamino)ethyl)amino)pyridine", "Ethylenediamine, N-(p-chlorobenzyl)-N',N'-dimethyl-N-(2-pyridyl)-", "N'-[(4-chlorophenyl)methyl]-N,N-dimethyl-N'-pyridin-2-ylethane-1,2-diamine", "Pyridine, 2-((p-chlorobenzyl)(2-(dimethylamino)ethyl)amino)-", "1,2-Ethanediamine, N-[(4-chlorophenyl)methyl]-N',N'-dimethyl-N-2-pyridinyl-", "p-Chlorbenzyl-.alpha.-pyridyl-dimethyl-aethylendiamin", "2-[(p-Chlorobenzyl)[2-(dimethylamino)ethyl]amino]pyridine", "Chloropyraminum", "Cloropiramina", "N-(4-Chlorobenzyl)-N',N'-dimethyl-N-(pyridin-2-yl)ethane-1,2-diamine", "Chloropyraminum [INN-Latin]", "Cloropiramina [INN-Spanish]", "NCGC00016672-01", "NCGC00016672-04", "EINECS 200-421-1", "CAS-6170-42-9", "BRN 0257085", "Pyridine, 2-[(p-chlorobenzyl)[2-(dimethylamino)ethyl]amino]-", "Chloropyramine (INN)", "p-Chlorbenzyl-alpha-pyridyl-dimethyl-aethylendiamin [German]", "N-((4-CHLOROPHENYL)METHYL)-N',N'-DIMETHYL-N-2-PYRIDINYL-1,2-ETHANEDIAMINE", "Prestwick0_000807", "Prestwick1_000807", "Prestwick2_000807", "Prestwick3_000807", "SCHEMBL30021", "BSPBio_000673", "p-Chlorbenzyl-alpha-pyridyl-dimethyl-aethylendiamin", "SPBio_002594", "BPBio1_000741", "2-((p-Chlorobenzyl)(2-dimethylaminoethyl)amino)pyridine", "CHEMBL1194287", "CHEBI:94767", "DTXSID50207729", "1H-Indole, 2-(2-bromophenyl)-", "NSC241038", "STL483421", "ZINC20148995", "AKOS024323312", "WLN: T6NJ BN2N1&1&1R DG", "DB08800", "MCULE-5568455528", "NSC-241038", "NCGC00016672-02", "NCGC00016672-07", "SBI-0207051.P001", "D07195", "170C429", "Q729863", "2-[(p-Chlorobenzyl)[2-(dimethylamino)ethyl]pyridine", "BRD-K83257731-003-03-7", "BRD-K83257731-003-14-4", "1, N-[(4-chlorophenyl)methyl]-N',N'-dimethyl-N-2-pyridinyl-", "2-[(4-CHLOROBENZYL)[2-(DIMETHYLAMINO)ETHYL]AMINO]PYRIDINE", "N,N-Dimethyl-N'-(p-chlorobenzyl)-N'-(2-pyridyl)ethylenediamine", "N-[(4-chlorophenyl)methyl]-N-[2-(dimethylamino)ethyl]pyridin-2-amine", "1,2-Ethanediamine, N-((4-chlorophenyl)methyl)-N',N'-dimethyl-N-2-pyridinyl- (9CI)", "N'-[(4-chlorophenyl)methyl]-N,N-dimethyl-N'-(2-pyridinyl)ethane-1,2-diamine", "N\'-[(4-chlorophenyl)methyl]-N,N-dimethyl-N\'-pyridin-2-ylethane-1,2-diamine"]" in row "1349" and field "synonyms" at position "4": type is "array/default"

**Comment by abradyIGS** [comment]
@jeremyjyang i'm moving this last issue to the helpdesk; as soon as i can figure out how to add you as a watcher, you should get an email

@jeevangelista after some further reading, i think i have to tweak one more thing to process these [censored] backslashes properly; stand by for another fix attempt, and sorry for confusion in earlier replies

**Comment by abradyIGS** [comment, marked as answer]
[FYI: I tried to port this to JIRA and I think I failed, but if anyone can actually access the ticket [here](https://cfde.atlassian.net/browse/HELP-181), please let me know: it would help with internal access-management debugging]

@jeremyjyang @jeevangelista **Please download the latest prep script, compound and substance tables from [here](https://osf.io/bq6k9/):**  if either of your respective issues remains unsolved by this update, please attach your submission TSVs [somewhere on github I can get to?] for immediate followup (and email support@cfde.atlassian.net so I can fold further developments more effectively into our issue tracker).

PS. Also be sure to grab the latest [JSON schema file](https://osf.io/c63aw/) for inclusion with your submission; we removed some uniqueness constraints (relevant to @jeremyjyang at minimum) on 4/8.
